src/main/java/org/gaul/s3proxy/S3ProxyHandler.java

```diff
@@ -2365,21 +2371,48 @@ private void handleUploadPart(HttpServletRequest request,
         }
     }
 
-    private static void addMetadataToResponse(HttpServletResponse response,
+    private static void addResponseHeaderWithOverride(
+            HttpServletRequest request,
+            HttpServletResponse response,
+            String headerName,
+            String overrideHeaderName,
+            String value) {
+        String override = request.getParameter(overrideHeaderName);
+        response.addHeader(headerName, override != null ? override : value);
```

Comment:
Note that while trying to use `Objects.firstNonNull` I got lots of `NullPointerException` in tests. Probably due to not having a fallback value when calling `Objects.firstNonNull`, probably because some object metadata values were not set.
Instead I used the ternary operator.